### PR TITLE
[CELEBORN-1202][FOLLOWUP] Update LICENSE and NOTICE files

### PR DESCRIPTION
**Pull request by pan3793**
### What changes were proposed in this pull request?

Update LICENSE and NOTICE files according to the mailing list comments.

### Why are the changes needed?

https://lists.apache.org/thread/zw5cw621dqgbktdolx7qynho0zt451pk

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Review.

### CODE REVIEW

File: NOTICE-binary [hunk 1]
```diff
@@ -9,6 +9,15 @@ The Apache Software Foundation (https://www.apache.org/).
...
 =============================================================================
 
+Apache Kyuubi
+Copyright 2021-2023 The Apache Software Foundation
+
+Remote Shuffle Service for Flink
+Copyright 2021 The Flink Remote Shuffle Project
+
+This product includes software developed at The Alibaba Group Holding Ltd
+(https://www.alibaba.com/).
```
Comment (pan3793) [inline NOTICE-binary:19]: the binary artifact contains all of Celeborn's source compiling outputs, so the content in source NOTICE should be present in binary NOTICE too

File: common/src/main/java/org/apache/celeborn/reflect/DynClasses.java [hunk 2]
```diff
@@ -22,7 +22,7 @@
 import java.util.LinkedHashSet;
 import java.util.Set;
 
-/** Copied from iceberg-common */
+/** this class is copied from Apache Kyuubi, derived from iceberg-common */
```
Comment (pan3793) [inline common/src/main/java/org/apache/celeborn/reflect/DynClasses.java:25]: update the comment to clarify the origin of this code

File: NOTICE-binary [hunk 3]
```diff
@@ -66,15 +78,9 @@ Copyright 2017-2020 The Apache Software Foundation
 Apache Ratis Thirdparty Miscellaneous
 Copyright 2017-2020 The Apache Software Foundation
 
-Apache Spark
-Copyright 2014 and onwards The Apache Software Foundation
```
Comment (pan3793) [inline NOTICE-binary:70]: moved to line 12

File: NOTICE-binary [hunk 4]
```diff
@@ -66,15 +78,9 @@ Copyright 2017-2020 The Apache Software Foundation
...
-
 ResolverUtil.java
 Copyright 2005-2006 Tim Fennell
 
-Apache Flink
-Copyright 2014-2022 The Apache Software Foundation
```
Comment (pan3793) [inline NOTICE-binary:76]: it's a mistake, we neither copied code from Flink nor ship Flink binary artifacts in the binary release

**Comment by pan3793** [comment]
@justinmclean, I have an additional question on writing the NOTICE file when including other TLP ASF projects.

According to https://infra.apache.org/licensing-howto.html#mod-notice

> It is important to keep NOTICE as brief and simple as possible, as each addition places a burden on downstream consumers.
>
> Do not add anything to NOTICE which is not legally required.
>
> If the dependency supplies a NOTICE file, its contents must be analyzed and the relevant portions bubbled up into the top-level NOTICE file.

As each ASF project's NOTICE file contains the following sentence

> This product includes software developed at
> The Apache Software Foundation (http://www.apache.org/).

I suppose it already indicates it includes ASF software, does it mean I can eliminate listing each ASF project included in this product?

The example NOTICE file listed on https://infra.apache.org/licensing-howto.html#example-notice is quite simple, I'm not sure why it does not mention other ASF products, because it does not include other ASF products, or just eliminate listing them?

**Comment by justinmclean** [comment]
No, you cannot remove each ASF project mentioned. If you include another ALv2 product with a NOTICE file, you must include the relevant it's of that NOTICE file in yours.

**Comment by justinmclean** [comment]
See the second part of https://infra.apache.org/licensing-howto.html#alv2-dep

**Comment by pan3793** [comment]
@justinmclean Thanks for your clarification, I assume that your point is "for projects under ALv2, there are no differences between non-ASF and ASF projects on NOTICE propagation"

Based on that, I updated the NOTICE to mention Parquet MR and Iceberg. Also opened https://github.com/apache/kyuubi/pull/5953 to fix the Apache Kyuubi side issue.

Also, would appreciate it if you could replace the NOTICE example https://infra.apache.org/licensing-howto.html#example-notice with a little bit of complex case.

**Comment by justinmclean** [comment]
There are no differences between non-ASF and ASF projects regarding the NOTICE files. BTW, I don't think you have included everything you need to in NOTICE; it is usually more than the project name and copyright line.

**Comment by pan3793** [comment]
@justinmclean we just copied a few classes from those projects, I don't think the reminding content in the NOTICE should be propagated

**Comment by justinmclean** [comment]
In some case IMO it needs to

**Comment by pan3793** [comment]
@justinmclean let me clarify with more details

According to https://infra.apache.org/licensing-howto.html

> In any case, the principle remains the same: LICENSE and NOTICE must exactly represent the contents of the distribution they reside in.

When coping code from an ALv2 project, I would check the LICENSE and the code-file comment, if no explicit declaration, I suppose the code is original from that project, in such case, I assume propagate the "project name and copyright line" of the NOTICE is sufficient.

But if there are clues indicate that the code was derived from other projects, give the `DynMethods` as an example, I copied it from Apache Kyuubi

```patch
- /** Copied from parquet-common */
+ /** this class is copied from Apache Kyuubi, derived from parquet-common */
public class DynMethods {
```
I would modify the comments to reflect the code derivation, and also analysis the content of Kyuubi's NOTICE, copy the related part of Parquet MR to the top-level NOTICE file.

**Comment by justinmclean** [comment]
Assuming propagating only the "project name and copyright line" of the NOTICE is sufficient is not correct. All lines of the NOTICE file should be considered for inclusion. Let us look at this one https://github.com/apache/parquet-mr/blob/master/NOTICE

Parts that should be copied IMO include:
- project name and copyright line
- Do any of the files copied contain code from parquet-tools, parquet-protobuf, Apache Avro, Kite or Netflix? It's quite hard to determine that.

So projects take two options: some will include the whole NOTICE file so they are covered, and others will check the history of each file included and work that out.

**Comment by pan3793** [comment]
> So projects take two options: ... others will check the history of each file included and work that out.

I checked https://github.com/apache/parquet-mr/pull/384/commits/d71836347cbe84e99fad7166fe89e99abf1b51ae, the original `DynMethods` should be imported from Cloudera/Kite, will check others and update NOTICE accordingly.

**Comment by pan3793** [comment]
@justinmclean I checked and updated the NOTICE. The other project's NOTICE files are easy to analyze, and I don't find any additional confusing parts. PTAL, thanks for your patience and detailed guidance.

**Comment by pan3793** [comment]
I'm going to merge this PR as there are no further comments.

**Comment by pan3793** [comment]
Merged to main/0.4

**Pull request closed without merging**